### ISSUE TITLE
[FEA] Add NVTX ranges to pool allocation/deallocation

### ISSUE THREAD

**Issue by rongou** [rapidsai/rmm#495]
**Is your feature request related to a problem? Please describe.**
It would be helpful to be able to see on an Nsight profile how much time is spend on allocating/deallocating memory in the pool memory resource, especially in a multi-threaded environment with per-thread default stream.

**Describe the solution you'd like**
Add NVTX ranges to memory allocation/deallocation in the pool.

**Additional context**
cuDF has a macro defined: https://github.com/rapidsai/cudf/blob/branch-0.15/cpp/include/cudf/detail/nvtx/ranges.hpp

**Comment by jrhemstad** [comment]
@harrism went about this [in the past](https://github.com/rapidsai/rmm/pull/336). We ended up not going through with it because most allocation events are faster than the recommended 1us minimum time for events to annotate with NVTX. That said I don't see anything wrong with adding it as an option that is disabled by default.

I think we can take a simpler/coarser grained approach than https://github.com/rapidsai/rmm/pull/336 and just annotate the `device_memory_resource` base class `allocate/deallocate` calls, that way we can see the annotations no matter what resource is being used.

**Comment by jrhemstad** [comment]
We're likely going to run into some difficulty/conflicts with having the `nvtx3.hpp` header in both `RMM` and `libcudf` until https://github.com/NVIDIA/NVTX/pull/2 is merged and we can pull the header from there to ensure both have the same version of the header.

**Comment by rongou** [comment]
There is some concern with contentions around CUDA events used in the pool between different threads/streams, it'd be nice to have better insight into that scenario.

**Comment by harrism** [comment]
I have an open PR #336 but I am waiting (as Jake points out) on NVIDIA/NVTX#2 and I will just add NVTX regions at the top-level rather than at low levels within `pool_memory_resource`.

**Comment by harrism** [comment]
Still waiting on NVIDIA/NVTX#2
